### PR TITLE
[PyTorch] Reduce CPU overhead on eager EP dispatch and combine

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description

 Lowers the per-call CPU cost of the eager NCCL EP dispatch and combine autograd paths. Numerics and public API are unchanged. Eager ops keep the CPU near the critical path (the dispatch forward must read the per-step recv count on the host before sizing recv), so Python and dispatch overhead was exposed rather than hidden behind async GPU work.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

  - Fuse eager prepare and dispatch into one C++ op and unify the eager and non-eager paths through it. The recv-count read (a host sync on pinned UVA memory, no D2H copy) and the recv-output allocation now happen inside the C++ op, so no Python runs between the count read and the dispatch launch.
  - Shorter autograd operand lists: `_EpPrepareAndDispatch` and `_EpCombine` take only the differentiable tensors as direct operands. The non-differentiable buffer tensors ride on the `EpBuffer` object, and `handle_mem` is stashed as a plain `ctx` attribute instead of `save_for_backward`.
  - Eager bypass of `torch.library`: eager is not graph-capturable, so combine forward, combine backward, and dispatch backward call the backend ops directly. Non-eager paths keep the `transformer_engine_ep::*` custom ops unchanged.
  - Cache `ep_get_zero_copy()` on `ctx` instead of re-querying it in every combine backward.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/ep.py [hunk 1]
```diff
@@ -483,125 +520,160 @@ def _ep_combine_raw(buffer: "EpBuffer", expert_out: torch.Tensor, result: torch.
...
                 recv_scale_inv,
-                token_counts,
+                tokens_per_expert,
                 tokens._fp8_dtype,
                 tokens.dtype,
             )
-        else:
-            recv_out = recv_tokens.detach()
-        return recv_out, recv_topk_weights.detach()
+            return recv_out, recv_topk_weights.detach()
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/ep.py]: We don’t have to call detach() for the eager case since we are constructing the tensor inside this autograd forward with requires_grad = False already. (For reducing CPU overheads in eager case). Becase AFAIK, detach crosses python to cpp boundary

File: transformer_engine/pytorch/ep.py [hunk 2]
```diff
@@ -611,33 +683,45 @@ class _EpCombine(torch.autograd.Function):
...
+        graph-capturable, so it calls the backend op directly and skips the torch.library dispatch.
+        """
+        handle_mem = buffer.handle_mem
+        token_counts = buffer.tokens_per_expert
+        eager = buffer.eager
         device = expert_out.device
-        result = torch.empty(num_local_tokens, hidden_dim, dtype=expert_out.dtype, device=device)
-        torch.ops.transformer_engine_ep.combine(handle_mem, expert_out, result)
-        ctx.save_for_backward(handle_mem)
+        zero_copy = tex.ep_get_zero_copy()
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/ep.py]: Can we cache this during  Bootstrap and reuse it. Since this is also a CPU overhead. The other option is we do the staging to symmetric memory in CPP itself?

**Review by vthumbe1503 — DISMISSED** [review]
LGTM

File: transformer_engine/pytorch/ep.py [hunk 3]
```diff
@@ -488,39 +526,76 @@ def _ep_combine_raw(buffer: "EpBuffer", expert_out: torch.Tensor, result: torch.
 # autograd.Function wrappers
 
 
-class _EpDispatch(torch.autograd.Function):
-    """Autograd dispatch; caller runs prepare first. bwd uses user-supplied grad inputs as-is."""
+@dataclass
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/ep.py]: Minor comment: using slots=True might be better for CPU overheads

Comment (phu0ngng) [reply transformer_engine/pytorch/ep.py]: Implemented.

**Review by vthumbe1503 — APPROVED** [review]
